File: Python3/프로그래머스/2/92342. 양궁대회/양궁대회.py
def solution(n, info):
    ryan = [0] * 11
    best_diff = 0
    answer = [-1]
    
    def get_diff(ryan, appeach):
        diff = 0
        for i in range(10):
            if ryan[i] > appeach[i]:
                diff += (10 - i)
            elif appeach[i] != 0:
                diff -= (10 - i)
        return diff
    
    def dfs(idx, used):
        nonlocal best_diff, answer
        if idx == 11 or used == n:
            ryan[10] += (n - used)
            
            diff = get_diff(ryan, info)
            if best_diff < diff:
                best_diff = diff
                answer = ryan[:]
            elif best_diff == diff and diff != 0:
                for i in range(10, -1, -1):
                    if answer[i] > ryan[i]:
                        break
                    elif answer[i] < ryan[i]:
                        answer = ryan[:]
                        break
                        
            ryan[10] -= (n - used)
            return
        
        needed = info[idx] + 1
        if (used + needed <= n):
            ryan[idx] = needed
            dfs(idx + 1, used + needed)
            ryan[idx] = 0
            
        dfs(idx + 1, used)
    
    dfs(0, 0)
    return answer

File: C++/프로그래머스/2/92342. 양궁대회/양궁대회.cpp
#include <string>
#include <vector>

using namespace std;

vector<int> answer = {-1};
vector<int> Info;
int best_score = 0;

int get_score(const vector<int>& cur) {
    int ryan = 0, apeach = 0;
    for (int i = 0; i < cur.size(); ++i) {
        if (cur[i] == 0 && Info[i] == 0) continue;
        if (cur[i] > Info[i]) ryan += (10 - i);
        else apeach += (10 - i);
    }
    return ryan - apeach;
}

bool is_better(const vector<int>& a, const vector<int>& b) {
    for (int i = 10; i >= 0; --i) {
        if (a[i] != b[i]) return a[i] > b[i];
    }
}

void dfs(int N, int pos, vector<int> cur) {
    if (N == 0 || pos == 11) {
        if (N > 0) cur[10] += N;
        int diff = get_score(cur);
        if (diff > 0) {
            if (diff > best_score) {
                best_score = diff;
                answer = cur;
            }
            else if (diff == best_score && is_better(cur, answer)) {
                answer = cur;
            }
        }
        if (N > 0) cur[10] -= N;
        return;
    }
    
    if (N > Info[pos]) {
        cur[pos] = Info[pos] + 1;
        dfs(N - cur[pos], pos + 1, cur);
        cur[pos] = 0;
    }
    dfs(N, pos + 1, cur);
}

vector<int> solution(int n, vector<int> info) {
    Info = info;
    vector<int> cur(info.size(), 0);
    dfs(n, 0, cur);
    
    return answer;
}